File: benchmarking/mpi_matmul.c
/* mpi_matmul.c */
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

int main(int argc, char **argv) {
    int N = 1024;       // Default matrix size (NxN)
    int count = 10;     // Number of timed iterations
    int warmup = 2;     // Number of warmup iterations (not measured)
    if (argc > 1) {
        N = atoi(argv[1]);
    }
    if (argc > 2) {
        count = atoi(argv[2]);
    }
    if (argc > 3) {
        warmup = atoi(argv[3]);
    }
    int total_iterations = count + warmup;

    MPI_Init(&argc, &argv);
    
    int rank, world_size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &world_size);
    
    // Allocate matrices.
    double *A = (double *) malloc(N * N * sizeof(double));
    double *B = (double *) malloc(N * N * sizeof(double));
    double *local_C = (double *) malloc(N * N * sizeof(double));
    double *C = (double *) malloc(N * N * sizeof(double));
    if (A == NULL || B == NULL || local_C == NULL || C == NULL) {
        fprintf(stderr, "Memory allocation failed on rank %d.\n", rank);
        MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
    }
    
    // Rank 0 initializes matrices A and B.
    if (rank == 0) {
        srand(time(NULL));
        for (int i = 0; i < N * N; i++){
            A[i] = (double) rand() / RAND_MAX;
            B[i] = (double) rand() / RAND_MAX;
        }
    }
    // Broadcast matrices.
    MPI_Bcast(A, N * N, MPI_DOUBLE, 0, MPI_COMM_WORLD);
    MPI_Bcast(B, N * N, MPI_DOUBLE, 0, MPI_COMM_WORLD);
    
    // Partition the inner (K) dimension among processes.
    int base = N / world_size;
    int rem = N % world_size;
    int k_start = rank * base + (rank < rem ? rank : rem);
    int k_count = base + (rank < rem ? 1 : 0);
    int k_end = k_start + k_count;
    
    double *timings = (double *) malloc(count * sizeof(double));
    if (timings == NULL) {
        fprintf(stderr, "Memory allocation for timings failed on rank %d.\n", rank);
        MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
    }
    int measured = 0;
    
    for (int iter = 0; iter < total_iterations; iter++){
        // Reset local_C to zero.
        for (int i = 0; i < N * N; i++){
            local_C[i] = 0.0;
        }
        
        double t0 = MPI_Wtime();
        
        //Is the implementation of torch matmul much more optimised than this which is causing network latency to overexceed this?

        for (int i = 0; i < N; i++){
            for (int k = k_start; k < k_end; k++){
                double a_val = A[i * N + k];
                for (int j = 0; j < N; j++){
                    local_C[i * N + j] += a_val * B[k * N + j];
                }
            }
        }
        
        // Sum up partial results.
        MPI_Allreduce(local_C, C, N * N, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
        
        double t1 = MPI_Wtime();
        if (iter >= warmup) {
            timings[measured++] = t1 - t0;
        }
    }
    
    // Compute statistics.
    double t_min = timings[0], t_max = timings[0], sum = 0.0;
    for (int i = 0; i < count; i++){
        if (timings[i] < t_min) t_min = timings[i];
        if (timings[i] > t_max) t_max = timings[i];
        sum += timings[i];
    }
    double t_avg = sum / count;
    double variance = 0.0;
    for (int i = 0; i < count; i++){
        double diff = timings[i] - t_avg;
        variance += diff * diff;
    }
    variance /= count;
    double stddev = (t_avg > 0.0) ? (sqrt(variance) / t_avg * 100) : 0.0;
    
    // Only rank 0 prints the CSV header and result.
    if (rank == 0) {
        printf("Implemendtation,CPU_Count,Matrix_Size,Iterations,t_min(s),t_max(s),t_avg(s),stddev(%%)\n");
        printf("MPI,%d,%d,%d,%.6f,%.6f,%.6f,%.2f\n",
               world_size, N, count, t_min, t_max, t_avg, stddev);
    }
    
    free(A);
    free(B);
    free(local_C);
    free(C);
    free(timings);
    
    MPI_Finalize();
    return 0;
}

File: benchmarking/mpi_matmul.py
#!/usr/bin/env python3
from mpi4py import MPI
import numpy as np
import argparse

def main():
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()

    parser = argparse.ArgumentParser(description="MPI Matrix Multiplication")
    parser.add_argument("--size", type=int, default=1000, help="Matrix size (NxN)")
    args = parser.parse_args()
    N = args.size

    # Process 0 initializes matrices; others prepare empty arrays.
    if rank == 0:
        A = np.random.rand(N, N)
        B = np.random.rand(N, N)
    else:
        A = np.empty((N, N), dtype='d')
        B = np.empty((N, N), dtype='d')
    
    # Broadcast matrices A and B to all processes.
    comm.Bcast(A, root=0)
    comm.Bcast(B, root=0)

    # Partition the inner dimension (summation index) among processes.
    # Create a list of counts that divides N as evenly as possible.
    counts = [(N // size) + (1 if i < (N % size) else 0) for i in range(size)]
    displs = [sum(counts[:i]) for i in range(size)]
    k_start = displs[rank]
    k_end = k_start + counts[rank]

    # Each process computes its partial contribution:
    # C_partial = A[:, k_start:k_end] dot B[k_start:k_end, :]
    local_C = np.dot(A[:, k_start:k_end], B[k_start:k_end, :])

    # Allocate space for the final result.
    C = np.empty((N, N), dtype='d')
    # Allreduce to sum up the partial results from all processes.
    comm.Allreduce(local_C, C, op=MPI.SUM)

    if rank == 0:
        print("Result matrix shape:", C.shape)

if __name__ == "__main__":
    main()
